<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [53]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [54]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [24]:
# текст запроса
query_3_1 = f'''select count(v.id)
                from public.vacancies v
            '''

In [25]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
vacancies_count = pd.read_sql_query(query_3_1, connection)

print(vacancies_count)

   count
0  49197


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_36306/328751920.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [26]:
# текст запроса
query_3_1 = f'''select count(e.id)
                from public.EMPLOYERS e
            '''

In [27]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
employers_count = pd.read_sql_query(query_3_1, connection)
# не забываем закрыть соединение после окончания работы

print(employers_count)

   count
0  23501


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_36306/3139747732.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_count = pd.read_sql_query(query_3_1, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [28]:
# текст запроса
query_3_1 = f'''select count(a.id)
                from public.areas a
            '''

In [29]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
areas_count = pd.read_sql_query(query_3_1, connection)

print(areas_count)

   count
0   1362


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_36306/3262544708.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_count = pd.read_sql_query(query_3_1, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [30]:
# текст запроса
query_3_1 = f'''select count(i.id)
                from public.industries i
            '''

In [31]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
industries_count = pd.read_sql_query(query_3_1, connection)
# не забываем закрыть соединение после окончания работы
print(industries_count)

# не забываем закрыть соединение после окончания работы
connection.close()

   count
0    294


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_36306/1372860772.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries_count = pd.read_sql_query(query_3_1, connection)


***

In [ ]:
# выводы по предварительному анализу данных

В среднем на одного работадателя приходится чуть больше 2 вакансий по количеству. Регионов представлено много (1362). Сфер деятельности 294.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''select a.name area, count(v.id) cnt
                from public.vacancies v
                    left join public.areas a on v.area_id = a.id
                group by a.id
                order by cnt desc
            '''

In [13]:
# результат запроса

# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
vacancies_count_by_area = pd.read_sql_query(query_4_1, connection)
# не забываем закрыть соединение после окончания работы
print(vacancies_count_by_area)

                         area   cnt
0                      Москва  5333
1             Санкт-Петербург  2851
2                       Минск  2112
3                 Новосибирск  2006
4                      Алматы  1892
..                        ...   ...
764                    Кизляр     1
765                    Джизак     1
766                    Эртиль     1
767                  Арсеньев     1
768  Тура (Красноярский край)     1

[769 rows x 2 columns]


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/2102520952.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count_by_area = pd.read_sql_query(query_4_1, connection)


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [17]:
# текст запроса
query_4_2 = f'''SELECT count(v.id) cnt
                FROM public.vacancies v
                WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
            '''

In [19]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
vacancies_count_salary_not_null = pd.read_sql_query(query_4_2, connection)
# не забываем закрыть соединение после окончания работы
print(vacancies_count_salary_not_null)

     cnt
0  24073


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/1872390566.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count_salary_not_null = pd.read_sql_query(query_4_2, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [22]:
# текст запроса
query_4_3 = f'''SELECT AVG(v.salary_from) avg_from, AVG(v.salary_to) avg_to
                FROM public.vacancies v
            '''

In [33]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
avg_salary_from_to = pd.read_sql_query(query_4_3, connection)
# не забываем закрыть соединение после окончания работы
print(round(avg_salary_from_to['avg_from']))
print(round(avg_salary_from_to['avg_to']))

0    71065.0
Name: avg_from, dtype: float64
0    110537.0
Name: avg_to, dtype: float64


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/1760940859.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary_from_to = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [38]:
# текст запроса
query_4_4 = f'''SELECT v.schedule, v.employment, COUNT(v.id)
                FROM public.vacancies v
                GROUP BY v.schedule, v.employment
                ORDER BY COUNT(v.id) desc
            '''

In [39]:
# результат запроса
vacancies_by_work_schedule = pd.read_sql_query(query_4_4, connection)
print(vacancies_by_work_schedule)

            schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/2396343030.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_by_work_schedule = pd.read_sql_query(query_4_4, connection)


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [43]:
# текст запроса
query_4_5 = f'''SELECT v.experience, COUNT(v.id)
                FROM public.vacancies v
                GROUP BY v.experience
                ORDER BY COUNT(v.id)
            '''

In [44]:
# результат запроса
vacancies_by_experience = pd.read_sql_query(query_4_5, connection)
print(vacancies_by_experience)

           experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/2840842267.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_by_experience = pd.read_sql_query(query_4_5, connection)


***

In [ ]:
# выводы по детальному анализу вакансий
Все довольно предсказуемо. Наиболее востребованы сотрудники на полный рабочий день с полной занятостью. С опытом от 1 года до 3 лет. Средний уровень зарплаты находится в диапазоне от 71 до 110 тыс. руб. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [49]:
# текст запроса
query_5_1 = f'''SELECT e.name, COUNT(v.id)
                FROM public.vacancies v
                LEFT JOIN public.employers e ON v.employer_id = e.id
                GROUP BY e.id
                ORDER BY COUNT(v.id) desc
  
            '''

In [50]:
# результат запроса
employers_rating = pd.read_sql_query(query_5_1, connection)
print(employers_rating)

/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/383789935.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_rating = pd.read_sql_query(query_5_1, connection)


                              name  count
0                           Яндекс   1933
1                       Ростелеком    491
2                         Тинькофф    444
3                             СБЕР    428
4                    Газпром нефть    331
...                            ...    ...
14901   Евсеев Сергей Владимирович      1
14902          Кадровые технологии      1
14903                   ЛОМБАРД 24      1
14904            Авто-Компонент ТД      1
14905  НПП «Старт» им. А.И. Яскина      1

[14906 rows x 2 columns]


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [83]:
# текст запроса
query_5_2 = f'''SELECT a.name area_name, COUNT(e.id) employers_count, COUNT(v.id) vacancies_count
                FROM public.areas a
                LEFT JOIN public.vacancies v ON v.area_id = a.id
                LEFT JOIN public.employers e ON e.area = a.id
                GROUP BY a.id
                ORDER BY COUNT(v.id) ASC, COUNT(e.id) DESC
  
            '''


In [84]:
# результат запроса
areas_info = pd.read_sql_query(query_5_2, connection)
print(areas_info)

/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/2620487617.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_info = pd.read_sql_query(query_5_2, connection)


               area_name  employers_count  vacancies_count
0                 Россия              410                0
1              Казахстан              207                0
2     Московская область               75                0
3     Краснодарский край               19                0
4               Беларусь               18                0
...                  ...              ...              ...
1357         Новосибирск          1149438          1149438
1358              Алматы          1364132          1364132
1359               Минск          2354880          2354880
1360     Санкт-Петербург          6320667          6320667
1361              Москва         31272712         31272712

[1362 rows x 3 columns]


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [86]:
	
query_5_3 = f'''SELECT e.name, count(distinct v.area_id) areas_count
            FROM public.employers e 
                JOIN public.vacancies v ON e.id = v.employer_id
            GROUP BY e.id
            ORDER BY count(distinct v.area_id) DESC
            '''

In [87]:
# результат запроса
employer_areas = pd.read_sql_query(query_5_3, connection)
print(employer_areas)

/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/521403104.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_areas = pd.read_sql_query(query_5_3, connection)


                                         name  areas_count
0                                      Яндекс          181
1                                  Ростелеком          152
2                                  Спецремонт          116
3                      Поляков Денис Иванович           88
4                                   ООО ЕФИН            71
...                                       ...          ...
14901                            НПП Авиатрон            1
14902              Центр дистанционных торгов            1
14903  Городские Телекоммуникационные Системы            1
14904                       Введенский, Отель            1
14905              Группа Компаний Автоимпорт            1

[14906 rows x 2 columns]


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [91]:
# текст запроса
query_5_4 = f'''SELECT COUNT(*) employers_without_industries
            FROM public.employers e 
                LEFT JOIN public.employers_industries ei ON ei.employer_id = e.id
            WHERE ei.industry_id IS NULL
        ''' 

In [92]:
# результат запроса
employers_without_industries = pd.read_sql_query(query_5_4, connection)
print(employers_without_industries.iloc[0])

employers_without_industries    8419
Name: 0, dtype: int64


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/3270021487.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_without_industries = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [113]:
# текст запроса
query_5_5 = f'''SELECT e.name employer_name
            FROM public.employers_industries ei
                JOIN public.employers e ON ei.employer_id = e.id
            GROUP BY e.id
            HAVING COUNT(distinct ei.industry_id) = 4
            ORDER BY e.name
            OFFSET 2 LIMIT 1
        ''' 

In [114]:
# результат запроса
employer = pd.read_sql_query(query_5_5, connection)
print(employer)

  employer_name
0          2ГИС


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/945696271.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer = pd.read_sql_query(query_5_5, connection)


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [115]:
# текст запроса
query_5_6 = f'''SELECT COUNT(distinct ei.employer_id)
            FROM public.employers_industries ei
                JOIN public.industries i ON ei.industry_id = i.id
            WHERE i.name = 'Разработка программного обеспечения'
            ''' 

In [116]:
# результат запроса
sodtware_development_employers = pd.read_sql_query(query_5_6, connection)
print(sodtware_development_employers)

   count
0   3553


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/4112955255.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sodtware_development_employers = pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [119]:
# код для получения списка городов-милионников
million_city = (
    "Москва", "Санкт-Петербург", "Новосибирск", 
    "Екатеринбург", "Казань", "Нижний Новгород", 
    "Красноярск", "Челябинск", "Самара", "Уфа", 
    "Ростов-на-Дону", "Омск", "Краснодар", 
    "Воронеж", "Пермь", "Волгоград"
)

In [126]:
# текст запроса
# выводим информацию по городам-миллионникам и объединяем её с доп. строкой с общим количеством вакансий
query_5_7 = f'''(SELECT 
                a.name area_name,
                COUNT(v.id) vacancies_count
            FROM public.vacancies v
                JOIN public.employers e ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
            WHERE e.name = 'Яндекс' and a.name in {million_city}
            GROUP BY a.id
            ORDER BY COUNT(v.id) ASC)
            
            UNION ALL 

            (SELECT 
                'Total',
                COUNT (*) total_vacancies_count
            FROM public.vacancies v
                JOIN public.employers e ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
            WHERE e.name = 'Яндекс' and a.name in {million_city}) 
            
            ORDER BY vacancies_count           
           
        ''' 

In [127]:
# результат запроса
m_cities_stats = pd.read_sql_query(query_5_7, connection)
print(m_cities_stats)

          area_name  vacancies_count
0              Омск               21
1         Челябинск               22
2        Красноярск               23
3         Волгоград               24
4             Пермь               25
5    Ростов-на-Дону               25
6            Казань               25
7            Самара               26
8               Уфа               26
9         Краснодар               30
10          Воронеж               32
11      Новосибирск               35
12  Нижний Новгород               36
13     Екатеринбург               39
14  Санкт-Петербург               42
15           Москва               54
16            TOTAL              485


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/3926702749.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  m_cities_stats = pd.read_sql_query(query_5_7, connection)


***

In [128]:
# выводы по анализу работодателей

# Мы посчитали, сколько вакансий открыто у различных работодателей,
# для каждого региона вывели количество работодателей и вакансий в нем,
# для каждого работодателя посчитали количество регионов, в которых у него есть вакансии,
# определили количество работодателей, у которых сфера деятельности не указана, 
# а также нашли работодателей с определенным количеством/названиями сфер деятельности,
# вывели распределение количества вакансий конкретного работодателя по городам-миллионникам.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [129]:
# текст запроса
query_6_1 = f'''SELECT COUNT(id)
            FROM public.vacancies
            WHERE    
                lower(name) like '%data%'
                OR lower(name) like '%данн%'
            ''' 

In [130]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
print(df)

   count
0   1771


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/1309227290.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [141]:
# текст запроса
query_6_2 = f'''SELECT COUNT(id)
                FROM public.vacancies
                WHERE (lower(name) like '%data scientist%'
                OR lower(name) like '%data science%'
                OR lower(name) like '%исследователь данных%'
                OR (name like '%ML%' AND lower(name) not like '%html%')
                OR lower(name) like '%machine learning%' 
                OR lower(name) like '%машинн%обучен%')
                
                AND (lower(name) like '%junior%'
                   OR experience = 'Нет опыта'
                   OR employment = 'Стажировка')
            '''

In [142]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
print(df)

   count
0     51


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/2386638782.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [144]:
# текст запроса
query_6_3 = f'''SELECT COUNT(id)
                FROM public.vacancies
                WHERE (lower(name) like '%data scientist%'
                    OR lower(name) like '%data science%'
                    OR lower(name) like '%исследователь данных%'
                    OR (name like '%ML%' AND lower(name) not like '%html%')
                    OR lower(name) like '%machine learning%' 
                    OR lower(name) like '%машинн%обучен%')
                    AND (lower(key_skills) like '%sql%' 
                        OR lower(key_skills) like '%postgres%')
            '''

In [145]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
print(df)

   count
0    201


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/1689521060.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [148]:
# текст запроса

query_6_4 = f'''SELECT COUNT(id)
                FROM public.vacancies
                WHERE (lower(name) like '%data scientist%'
                    OR lower(name) like '%data science%'
                    OR lower(name) like '%исследователь данных%'
                    OR (name like '%ML%' AND lower(name) not like '%html%')
                    OR lower(name) like '%machine learning%' 
                    OR lower(name) like '%машинн%обучен%')
                    AND (key_skills like '%Python%')
            '''

In [149]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
print(df)

   count
0    351


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/1255022825.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [161]:
# текст запроса
query_6_5 = f'''SELECT 
                    ROUND(AVG(
                        length(key_skills) + 1 - length(REPLACE(key_skills, CHR(9), ''))
                    ), 2) 
                FROM public.vacancies 
                WHERE ((lower(name) like '%data scientist%'
                    OR lower(name) like '%data science%'
                    OR lower(name) like '%исследователь данных%'
                    OR (name like '%ML%' AND lower(name) not like '%html%')
                    OR lower(name) like '%machine learning%' 
                    OR lower(name) like '%машинн%обучен%'))
                    AND (key_skills IS NOT NULL)
            '''

In [162]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
print(df)

/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/4051561159.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


   round
0   6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [163]:
# текст запроса
query_6_6 = f'''SELECT 
                    experience, 
                    ROUND(AVG(
                        coalesce((salary_from+salary_to)/2, salary_from, salary_to)
                    ), 0)
                FROM public.vacancies 
                WHERE 
                    (((lower(name) like '%data scientist%'
                    OR lower(name) like '%data science%'
                    OR lower(name) like '%исследователь данных%'
                    OR (name like '%ML%' AND lower(name) not like '%html%')
                    OR lower(name) like '%machine learning%' 
                    OR lower(name) like '%машинн%обучен%')))
                    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                GROUP BY experience         
            ''' 

In [165]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
print(df)

           experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/1300537.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


***

In [ ]:
# выводы по предметному анализу

# Здесь нашли вакансии, связанные с данными, а также вакансии конкретно для дата-сайентистов.
# Среди последних посчитали количество тех, которые подходят начинающим специалистам, 
# а также количество вакансий, в которых в качестве ключевых навыков указаны SQL/Postrges или Python, 
# посчитали среднее количество требуемых ключевых навыков,
# определили среднюю зарплату в зависимости от опыта.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Наибольшее количество вакансий нашей базы данных находится в столицах и городах миллионниках, в небольших населённых пунктах вакансий гораздо меньше. Что собственно логично.

In [167]:
query = f'''SELECT 
                i.name,
                count(v.id)
            FROM 
                public.industries i
                JOIN public.employers_industries ei ON ei.industry_id = i.id
                JOIN public.employers e ON ei.employer_id = e.id
                JOIN public.vacancies v ON v.employer_id = e.id
            GROUP BY i.name 
            ORDER BY 2 DESC        
        ''' 
df = pd.read_sql_query(query, connection)
print(df)

                                                  name  count
0                  Разработка программного обеспечения  12499
1    Системная интеграция,  автоматизации технологи...  11034
2    Интернет-компания (поисковики, платежные систе...   6413
3                                                 Банк   2742
4                                      Мобильная связь   1585
..                                                 ...    ...
288                               Спортивная федерация      5
289                                          Химчистка      4
290             Ассоциация в сфере культуры, искусства      2
291                                          Автошкола      1
292                                  Ритуальные услуги      1

[293 rows x 2 columns]


/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/2381659707.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Как видно, максимальное количество вакансий находятся в таких сферах деятельности  как "Разработка программного обеспечения" и "Системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг". Дальше в рейтинге идут Интернет компании, Банки и Операторы сотовой связи. Не исключением по отраслям является и спрос на DS. А вот наименьшее количество вакансий в Автошколах и Ритуальных услугах. 


In [169]:
#ключевые навыки для новичков DS
query = f'''SELECT key_skills
            FROM public.vacancies v
            WHERE (lower(v.name) like '%data scientist%'
                OR lower(v.name) like '%data science%'
                OR lower(v.name) like '%исследователь данных%'
                OR (v.name like '%ML%' AND lower(v.name) not like '%html%')
                OR lower(v.name) like '%machine learning%' 
                OR lower(v.name) like '%машинн%обучен%')
                
                AND (lower(v.name) like '%junior%'
                OR v.experience = 'Нет опыта'
                OR v.employment = 'Стажировка')
            ORDER BY 1
        ''' 
df = pd.read_sql_query(query, connection)
print(df)

                                           key_skills
0   Computer Vision\tPyTorch\tTensorflow\tDeep Lea...
1                                       Linux\tPython
2                                       Linux\tPython
3   Machine Learning\tDeep Learning\tData Science\...
4   Machine Learning\tDeep Learning\tData Science\...
5   Machine Learning\tDeep Learning\tData Science\...
6   ML\tлинейные модели\tRandom Forest\tGradient B...
7   Python\tC++\tОптимизация кода\tАнализ кода\tМа...
8          Python\tDjango Framework\tSQL\tXML\tMS SQL
9   Python\tGit\tTensorFlow\tKeras\tJupiter\tNLP\t...
10  Python\tLinux\tGit\tSQL\tDocker\tBash\tOpenCV\...
11  Python\tMachine Learning\tPandas\tSQL\tNumpy\t...
12  Python\tMachine Learning\tМатематическое модел...
13  Python\tMATLAB\tМатематическое моделирование\t...
14                                         Python\tML
15               Python\tML\tPytorch\tComputer Vision
16  Python\tOpenCV\tTensorflow\tComputer Vision\tН...
17  Python\tPyTorch\tGit\tTe

/var/folders/9t/d4bq7zb146v439lxxk1zxns00000gq/T/ipykernel_67786/211770615.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Ключевой навык для DS это Python, Математика, SQL.

В дальнейшем можно проанализирувать другие профессии. 